app/Http/Controllers/TrainingDash.php

```diff
@@ -567,67 +573,366 @@ public function otsCancel($id) {
+        }
+        // Number of unique students
+        if ($dataType == 'stats') {
+            $uniqueStudents = $sessions->unique('controller_id')->count();
+            $retArr['uniqueStudents'] = $uniqueStudents;
+        }
+        // Number of OTS' per month
+        if ($dataType == 'stats') {
+            $otsMonth = Ots::whereBetween('updated_at', [$from . ' 00:00:00', $to . ' 00:00:00'])->get();
+            $otsPerMonthPass = $otsMonth->where('status', '2')->count();
+            $otsPerMonthFail = $otsMonth->where('status', '3')->count();
+            $retArr['otsPerMonth'] = ['pass' => $otsPerMonthPass, 'fail' => $otsPerMonthFail];
+        }
+        // Average session duration (over last 6-months)
+        if ($dataType == 'graph') {
+            $sessionsSixMonths = TrainingTicket::where('start_date', '>', Carbon::now()->subMonths(6))->get();
+            $uniqueSessionIDs = $sessionsSixMonths->unique('session_id');
+            $sessionDuration = [];
+            foreach ($uniqueSessionIDs as $uniqueSessionID) {
+                $seshByTypeDuration = $averageSeshDuration = null;
+                $seshByType = $sessionsSixMonths->where('session_id', $uniqueSessionID->session_id);
+                foreach ($seshByType as $sesh) {
+                    $seshByTypeDuration += strtotime("1970-01-01 " . $sesh->duration . ":00");
+                }
+                $averageSeshDuration = $seshByTypeDuration / $seshByType->count() / 60; // Get duration in minutes
+                $sessionDuration[] = [$uniqueSessionID->getSessionNameAttribute(), round($averageSeshDuration, 2)];
+            }
+            $retArr['sessionDuration'] = $sessionDuration;
+        }
+        // Number of student cancellations, no shows
+        if ($dataType == 'stats') {
+            $setmoreAccessToken = $setmoreCursor = null;
+            $setmoreAppointments = [];
+            $studentCancel = $studentNoShow = 0;
+            $client = new Client();
+            try {
+                $res = $client->request('GET', Config::get('setmore.endpoint') . '/o/oauth2/token?refreshToken=' . Config::get('setmore.api_key'));
+                $setmoreAuthResp = (string) $res->getBody();
+                $setmoreAuthRespA = json_decode($setmoreAuthResp, true);
+            } catch (ClientErrorResponseException $exception) {
+                // Unused - don't throw an error
+            }
+            if (isset($setmoreAuthRespA['response'])) {
+                if ($setmoreAuthRespA['response'] && isset($setmoreAuthRespA['data']['token'])) {
+                    $setmoreAccessToken = $setmoreAuthRespA['data']['token']['access_token'];
+                }
+            }
+            if ($setmoreAccessToken != null) {
+                while ($this->getAppointmentBatch($setmoreAppointments, $setmoreAccessToken, $from, $to, $setmoreCursor)) {
+                }
```

Comment:
Using pointers can be sketchy... It would probably be better to move this into its own function and return the information necessary. It also makes this easier to read because an empty while loop looks strange.